### PR TITLE
[wptserve] Eliminate race condition

### PR DESCRIPTION
**Pull request by jugglinmike**
This race condition was expressed during testing sessions where the
first test to use the Stash feature issued did so with multiple requests
made in parallel.

---

The tests are brittle, but given the conditions we need to model, brittleness seems unavoidable.

The server code hasn't changed recently, so it's a little surprising that a race condition should suddenly become an issue [in Taskcluster](https://github.com/web-platform-tests/wpt/issues/13989) and [in the results-collection project](https://github.com/web-platform-tests/results-collection/issues/625).  Reviewing the status for each commit in `master` shows that [this intermittent error first surfaced on November 5](https://github.com/web-platform-tests/wpt/commits/master?after=7c568fdbd98ccdb92152c14521d3cc8c729a45a2+74).  Guy Fawkes is the obvious suspect, but the trail's gone cold. I turned to more technical explanations.

[The most recent Chromedriver release is version 2.43](https://chromedriver.storage.googleapis.com/index.html), and [that was published on October 16](https://chromedriver.storage.googleapis.com/2.43/notes.txt), ruling out a regression there. Chrome itself may have regressed, but I'm not set up to do any sort of bisecting, so I'm proceeding under the assumption that this is an issue in WPT.

A small incongruity between the Buildbot and Taskcluster setup is useful here.  Each system experiences the issue on a different "chunk" of WPT, but they define the segments in different terms. Buildbot uses 20 "chunks" containing tests of all types (failing on number 3) while Taskcluster uses 15 "chunks" for the testharness.js tests (failing on 13). We can narrow the set of suspect files by using the union of the following two lists:

    ./wpt run --list-tests --this-chunk 3 --total-chunks 20 chrome
    ./wpt run --list-tests --this-chunk 13 --total chunks 15 --test-type testdriver -- chrome

That returns 187 test files to consider. Feeding those files into `git log` shows that only one of them changed on that day:

    $ cat both.txt | xargs git log --after=2018-11-04 --before=2018-11-06 --oneline --
    7e66d13 Ensure eval flag is properly transfered to context from CSPRO

[`content-security-policy/script-src/eval-allowed-in-report-only-mode-and-sends-report.html` doesn't appear to be doing anything invalid.](https://github.com/web-platform-tests/wpt/blob/89188af97522b15cfe7ab6cc1c9cdb58007f0967/content-security-policy/script-src/eval-allowed-in-report-only-mode-and-sends-report.html) It does make parallel requests to a Stash-enabled endpoint, though (via [`checkReport.sub.js`](https://github.com/web-platform-tests/wpt/blob/89188af97522b15cfe7ab6cc1c9cdb58007f0967/content-security-policy/support/checkReport.sub.js)). The race condition fixed by this patch would only be expressed when the first test to use Stash did so via two parallel requests; my guess is that's uncommon but became true on November 5.

### CODE REVIEW

**Comment by gsnedders** [comment]
At first glance this looks good, but I definitely need to have a closer look while more awake; thanks for tracking this down!

**Review by jgraham — APPROVED** [review]
Excellent job diagnosing this issue and providing a fix.

I am slightly concerned about the tests; it took a lot longer to understand the test code than it did to understand the fix. If we want to keep them I think we should factor out the 90% common code, and comment it thoroughly so that it's easier to understand what is supposed to be happening at each point; a test involving two processes and two threads is inherently complex and difficult to understand.

File: tools/wptserve/tests/test_stash.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    yield add
+
+    for fn in fns:
+        fn()
+
+def test_delayed_lock(add_cleanup):
+    """Ensure that delays in proxied Lock retrieval do not interfere with
+    initialization in parallel threads."""
+
+    class SlowLock(BaseManager):
```
Comment (gsnedders) [inline tools/wptserve/tests/test_stash.py]: What makes this slow?

File: tools/wptserve/tests/test_stash.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    add_cleanup(lambda: parallel.terminate())
+
+    assert [queue.get(), queue.get()] == [False, False], (
+        "both instances had valid locks")
+
+def test_delayed_dict(add_cleanup):
+    """Ensure that delays in proxied `dict` retrieval do not interfere with
+    initialization in parallel threads."""
+
+    class SlowDict(BaseManager):
```
Comment (gsnedders) [inline tools/wptserve/tests/test_stash.py]: What makes this slow?

File: tools/wptserve/tests/test_stash.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    def handle_lock_request():
+        request_lock.release()
+        response_lock.acquire()
+        return threading.Lock()
+
+    SlowLock.register("get_dict", callable=lambda: {})
+    SlowLock.register("Lock", callable=handle_lock_request)
+
+    slowlock = SlowLock(("localhost", 4543), b"some key")
```
Comment (gsnedders) [inline tools/wptserve/tests/test_stash.py]: I hate to be stupid, but what effect does this manager have?

**Review by gsnedders — COMMENTED** [review]
(Mostly just agree with @jgraham)

**Comment by jugglinmike** [comment]
Thanks for the review, @jgraham and @gsnedders. I've attempted to address @jgraham's feedback by answering @gsnedders' questions via inline documentation. I've also reduced duplication by factoring out the multiprocessing target (`run`) and sharing it between both tests.

File: tools/wptserve/tests/test_stash.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    # Create a second thread in order to inspect the behavior of the Stash
+    # constructor at this moment.
+
+    second.start()
+
+    # Allow the `first` thread to proceed
+
+    response_lock.release()
+
+    # Wait for both threads to complete and report their stateto the test
```
Comment (jgraham) [inline tools/wptserve/tests/test_stash.py]: type: `stateto`

File: tools/wptserve/tests/test_stash.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    def handle_lock_request():
+        """This request handler allows the caller to delay execution of a
+        thread which has requested a proxied representation of the `lock`
+        property, simulating a "slow" interprocess communication channel."""
+
+        request_lock.release()
+        response_lock.acquire()
+        return threading.Lock()
+
+    SlowLock.register("get_dict", callable=lambda: {})
```
Comment (jgraham) [inline tools/wptserve/tests/test_stash.py]: It still somewhat seems like you could write these tets as as a parameterised test in which your parameters are `get_dict` and `Lock`, with one case having the parameters `(lambda :{}, mutex_lock_request)` and one having the parameters `(mutex_get_dict, lambda: threading.Lock())`. I also *think* the names with "mutex" in make more sense than the generic `handle_` names.

**Review by jgraham — APPROVED** [review]
This looks better, and I don't want to block it landing, but there is possibly scope for one more round of cleanup.

File: tools/wptserve/tests/test_stash.py (hunk 5 above)
Comment (jgraham) [reply tools/wptserve/tests/test_stash.py]: Or maybe `blockable` rather than `mutex`

**Comment by jugglinmike** [comment]
Those functions can't be parameterized like that because they need to reference
locks (currently named `request_lock` and `response_lock`) which are shared
with the child process created by the test.

We could also parameterize the locks, and maybe make some sort of
mutex-function-producing factory function, but that kind of abstraction is
contrary to our goal of making the test easier to understand.

I agree that `mutex_` is more descriptive than `handle_`. I've updated the
function names accordingly.

**Comment by jgraham** [comment]
I mean you could make the locks globals and set them at the start of the test and unset them in the cleanup. But I'm going to merge this PR now for the correctness fix, and if we want to clean up the tests more we can do that in a later PR.

**Comment by jugglinmike** [comment]
Thanks!

**Pull request merged**